ml4h/models.py

```diff
@@ -1273,6 +1722,157 @@ def make_paired_autoencoder_model(
+        tensor_maps_in: List[TensorMap],
+        tensor_maps_out: List[TensorMap],
+        encoder_blocks: List[str],
+        decoder_blocks: List[str],
+        merge_blocks: List[str],
+        learning_rate: float,
+        optimizer: str,
+        u_connect: DefaultDict[TensorMap, Set[TensorMap]] = None,
+        training_steps: int = None,
+        learning_rate_schedule: str = None,
+        **kwargs,
+) -> Tuple[Model, Dict[TensorMap, Model], Dict[TensorMap, Model]]:
+    """Make multi-task, multi-modal feed forward neural network for all kinds of prediction
+
+    This model factory can be used to make networks for classification, regression, and segmentation
+    The tasks attempted are given by the output TensorMaps.
+    The modalities and the first layers in the architecture are determined by the input TensorMaps.
+    Model architectures are specified by Blocks which can encode, merge or decode TensorMaps.
+
+    Hyperparameters are exposed to the command line and passed through to Block constructors via **kwargs.
+    Model summary printed to output
+
+    :param tensor_maps_in: List of input TensorMaps
+    :param tensor_maps_out: List of output TensorMaps
+    :param encoder_blocks: One or more BLOCK_CLASS keys or hd5 model files to encode applicable input TensorMaps
+    :param decoder_blocks: One or more BLOCK_CLASS keys or hd5 model files to decode applicable output TensorMaps
+    :param merge_blocks: Zero or more BLOCK_CLASS keys to construct multimodal latent space or apply loss functions based on internal activations, etc..
+    :param learning_rate: learning rate for optimizer
+    :param u_connect: dictionary of input TensorMap -> output TensorMaps to u connect to
+    :param optimizer: which optimizer to use. See optimizers.py.
+    :return: a compiled keras model
+    :param learning_rate_schedule: learning rate schedule to train with, e.g. triangular
+    :param training_steps: How many training steps to train the model. Only needed if learning_rate_schedule given
+    """
+    tensor_maps_out = parent_sort(tensor_maps_out)
+    u_connect: DefaultDict[TensorMap, Set[TensorMap]] = u_connect or defaultdict(set)
+    custom_dict = _get_custom_objects(tensor_maps_out)
+    opt = get_optimizer(
+        optimizer, learning_rate, steps_per_epoch=training_steps, learning_rate_schedule=learning_rate_schedule,
+        optimizer_kwargs=kwargs.get('optimizer_kwargs'),
+    )
+
+    if kwargs.get('model_file', False):
+        return _load_model_encoders_and_decoders(tensor_maps_in, tensor_maps_out, custom_dict, opt, kwargs['model_file'])
+
+    encoder_block_functions = {tm: identity for tm in tensor_maps_in}  # Dict[TensorMap, Block]
+    for tm in tensor_maps_in:
+        for encode_block in encoder_blocks:
+            if encode_block in BLOCK_CLASSES:
+                encoder_block_functions[tm] = compose(encoder_block_functions[tm], BLOCK_CLASSES[encode_block](tensor_map=tm, **kwargs))
```

Comment:
Does each `encode_block` get  applied to each input it `can_apply` to?
Is it possible to encode the scenario where I want to apply 2 dense blocks to an ECG and 1 dense block to an MRI?

Reply:
Yes,  but for now they would have to be serialized separately